### PR TITLE
add invalid propose metrics

### PR DESCRIPTION
**Pull request by sch00lb0y**
Signed-off-by: பாலாஜி ஜின்னா <rbalajis25@gmail.com>

## What have you changed? (mandatory)
I've added metrics for invalid raft proposal

and also changed RaftMetrics to RefCell RaftMetrics

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
- Summarize your change (mandatory)
```rust
  // Check store_id, make sure that the msg is dispatched to the right place.
        util::check_store_id(msg, self.store_id())?;
        if msg.has_status_request() {
            // For status commands, we handle it here directly.
            let resp = self.execute_status_command(msg)?;
            return Ok(Some(resp));
        }

        // Check whether the store has the right peer to handle the request.
        let peer = self.check_propose_peer(msg)?; //immutable borrow
        // peer_id must be the same as peer's.
        util::check_peer_id(msg, peer.peer_id())?;
        // Check whether the term is stale.
        util::check_term(msg, peer.term())?;

```
```let peer = self.check_propose_peer(msg)?;``` immutable borrow happening here so I've changed into RefCell for interior mutablility

Please **NOTE** that:
- Do not assume reviewers understand the original issue

## What are the type of the changes? (mandatory)
- Little bit refactoring and added invalid proposal metrics
- Improvement (non-breaking change which is an improvement to an existing feature)
## How has this PR been tested? (mandatory)
make dev
## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
no
## Does this PR affect tidb-ansible update? (mandatory)
not sure
If there is a configuration or metrics change, please file a PR in [tidb-ansible](https://github.com/pingcap/tidb-ansible), and add the PR number here.

## Refer to a related PR or issue link (optional)
#2613

### CODE REVIEW

**Comment by sch00lb0y** [comment]
I have updated the branch.

**Comment by siddontang** [comment]
There is only one place to call check_propose_peer, why not pass mut self to it?

**Comment by sch00lb0y** [comment]
@siddontang 
```rust
...
1503 |                 let sibling_region_id = self.find_sibling_region(peer.region());
     |                                         ^^^^ immutable borrow occurs here
...
1513 |     }
     |     - mutable borrow ends here

error[E0502]: cannot borrow `self.region_peers` as immutable because `*self` is also borrowed as mutable
    --> src/raftstore/store/fsm/peer.rs:1505:42
     |
1491 |         let peer = self.check_propose_peer(msg)?;
     |                    ---- mutable borrow occurs here
...
1505 |                     let sibling_region = self.region_peers[&sibling_region_id].region(); 
```
mutable borrow block is end of the function. can't able to do the immutable borrow in between.

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -1489,10 +1534,23 @@ impl<T: Transport, C: PdClient> Store<T, C> {
 
         // Check whether the store has the right peer to handle the request.
         let peer = self.check_propose_peer(msg)?;
```
Comment (overvenus) [inline src/raftstore/store/fsm/peer.rs]: How about inline the `check_propose_peer`? So that we borrow the `region_peers` instead of the `self`.

Comment (sch00lb0y) [reply src/raftstore/store/fsm/peer.rs]:
```rust
let region_id = msg.get_header().get_region_id();
        let peer = match self.region_peers.get(&region_id) {
            Some(peer) => peer,
            None => return Err(Error::RegionNotFound(region_id)),
        };
        if !peer.is_leader() {
            return Err(Error::NotLeader(
                region_id,
                peer.get_peer_from_cache(peer.leader_id()),
            ));
        }
```
Your saying to move this block to this fucntion right?

Comment (overvenus) [reply src/raftstore/store/fsm/peer.rs]: Yes, it looks like resolve borrow issue.

**Comment by sch00lb0y** [comment]
@overvenus I have updated it. :)

**Comment by overvenus** [comment]
PTAL @siddontang

**Comment by breeswish** [comment]
@overvenus Please add metrics to tidb-ansible. Thanks!

**Pull request merged**